add recipe for lem.el

mooseyboots opened the pull request:
### Brief summary of what the package does

emacs client for Lemmy, the federated link-aggregator.

### Direct link to the package repository
https://codeberg.org/martianh/lem.el

### Your association with the package
author

### Relevant communications with the upstream package maintainer

**None needed**

### Checklist

<!-- Please confirm by replacing `[]` with `[x]`: -->

- [X] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [X] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [X] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [X] My elisp byte-compiles cleanly
- [X] I've used `M-x checkdoc` to check the package's documentation strings
- [] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

<!-- After submitting, please fix any problems the CI reports. -->

I'm unable to build locally from my recipes, due to an error with straight.el that i don't understand. perhaps there's a way to build it outside of straight entirely?

mooseyboots commented:
fyi, the `hierarchy--make` that flycheck complains about is a struct constructor fun in hierarchy.el, which exists/is defined.

riscy commented:
Apologies for missing this for a couple weeks - I think my eye saw that it had been commented on once and I assumed it was a review comment.  Here's a quick first pass from my side --

`lem-api.el` with byte-compile using Emacs 28.1:
```
lem-api.el:350:7: Warning: Unused lexical variable `auth'
```
If it's truly unused, you can prepend with `_`

`lem-api.el` with checkdoc 0.6.2:
```
lem-api.el:187: Lisp symbol `fedi-package-prefix' should appear in quotes
```

`lem-api.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- lem-api.el#L140: Top-level `setq` should usually be replaced by `defvar`
- lem-api.el#L142: Top-level `setq` should usually be replaced by `defvar`
- lem-api.el#L145: Top-level `setq` should usually be replaced by `defvar`
```
Use `(defvar fedi-http--api-version "v3")` instead.

`lem-post.el` with byte-compile using Emacs 28.1:
```
lem-post.el:29:1: Error: Cannot open load file: No such file or directory, persist
```
This is missing a package-requires.

`lem-post.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- lem-post.el#L283: Loading a package should rarely add hooks
```
You shouldn't need to add advice here - you're fully in control of the mode definition.

`lem-search.el` with byte-compile using Emacs 28.1:
```
lem-search.el:50:6: Warning: the function `lem-get-community' is not known to
    be defined.
lem-search.el:47:48: Warning: the function `lem-search' is not known to be
    defined.
```

`lem-ui.el` with byte-compile using Emacs 28.1:
```
lem-ui.el:37:1: Error: Cannot open load file: No such file or directory, markdown-mode
```

`lem.el` with byte-compile using Emacs 28.1:
```
lem.el:31:1: Error: Cannot open load file: No such file or directory, emojify
```

`lem.el` with package-lint 20230806.1331:
```
1 issue found:
6:51: error: Package markdown is not installable.
```

`lem.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- The `;;; Commentary` for this file is much wider than 80 characters
```

`#'load`-check on each file:
```
Loading lem.el
  lem.el:Error: Emacs 28.1:
  (file-missing "Cannot open load file" "No such file or directory" "emojify")
Loading lem-ui.el
  lem-ui.el:Error: Emacs 28.1:
  (file-missing "Cannot open load file" "No such file or directory" "markdown-mode")
Loading lem-search.el
Loading lem-post.el
  lem-post.el:Error: Emacs 28.1:
  (file-missing "Cannot open load file" "No such file or directory" "persist")
Loading lem-api.el
```

mooseyboots commented:
> Apologies for missing this for a couple weeks - I think my eye saw that it had been commented on once and I assumed it was a review comment. Here's a quick first pass from my side --

no problem, and gosh i'm sorry about the mess! there are a few reasons for it, but one is that i added another file in the meantime and it's less clean. another is that my checking capabilities are clearly inferior to yours, for example regarding autoloads.

> `lem-api.el` with byte-compile using Emacs 28.1:
> 
> ```
> lem-api.el:350:7: Warning: Unused lexical variable `auth'
> ```
> 
> If it's truly unused, you can prepend with `_`

this was tricky, because in a macro, but i worked out how to prepend a `_` in the macro if auth remains undefined. thanks.

> `lem-api.el` with checkdoc 0.6.2:
> 
> ```
> lem-api.el:187: Lisp symbol `fedi-package-prefix' should appear in quotes
> ```

this seems wrong: it is given in example code. maybe i need to quote the code differently? i used `\=(`

> `lem-api.el` with [melpazoid](https://github.com/riscy/melpazoid):
> 
> ```
> - lem-api.el#L140: Top-level `setq` should usually be replaced by `defvar`
> - lem-api.el#L142: Top-level `setq` should usually be replaced by `defvar`
> - lem-api.el#L145: Top-level `setq` should usually be replaced by `defvar`
> ```
> 
> Use `(defvar fedi-http--api-version "v3")` instead.

here i'm setting vars from `fedi.el`, so if i just defvar them i get complaints about how they don't start with the pkg prefix. maybe there's a better way to deal with this in terms of organization?

> 
> `lem-post.el` with byte-compile using Emacs 28.1:
> 
> ```
> lem-post.el:29:1: Error: Cannot open load file: No such file or directory, persist
> ```
> 
> This is missing a package-requires.

i removed the code concerned from `fedi-post.el`, which is the source of this complaint. `persist` is currently unused in the libraries.

> 
> `lem-post.el` with [melpazoid](https://github.com/riscy/melpazoid):
> 
> ```
> - lem-post.el#L283: Loading a package should rarely add hooks
> ```
> 
> You shouldn't need to add advice here - you're fully in control of the mode definition.

i wish i had such good checking capabilities! mastodon.el also does this, i need to fix both. thanks.

> 
> `lem-search.el` with byte-compile using Emacs 28.1:
> 
> ```
> lem-search.el:50:6: Warning: the function `lem-get-community' is not known to
>     be defined.
> lem-search.el:47:48: Warning: the function `lem-search' is not known to be
>     defined.
> ```
fixed, thanks.

> `lem-ui.el` with byte-compile using Emacs 28.1:
> 
> ```
> lem-ui.el:37:1: Error: Cannot open load file: No such file or directory, markdown-mode
> ```
i guess this is because of erroneous package-requires in `lem.el`. its now changed to the correct `markdown-mode`.

> `lem.el` with byte-compile using Emacs 28.1:
> 
> ```
> lem.el:31:1: Error: Cannot open load file: No such file or directory, emojify
> ```
fixed with Package-Requires.

> `lem.el` with package-lint 20230806.1331:
> 
> ```
> 1 issue found:
> 6:51: error: Package markdown is not installable.
> ```
edited to `markdown-mode`, should be fixed.

> `lem.el` with [melpazoid](https://github.com/riscy/melpazoid):
> 
> ```
> - The `;;; Commentary` for this file is much wider than 80 characters
> ```
fixed, thanks.
 
> `#'load`-check on each file:
> 
> ```
> Loading lem.el
>   lem.el:Error: Emacs 28.1:
>   (file-missing "Cannot open load file" "No such file or directory" "emojify")

fixed. 

> Loading lem-ui.el
>   lem-ui.el:Error: Emacs 28.1:
>   (file-missing "Cannot open load file" "No such file or directory" "markdown-mode")

hopefully fixed by correcting the package-requires as mentioned above.

> Loading lem-search.el
> Loading lem-post.el
>   lem-post.el:Error: Emacs 28.1:
>   (file-missing "Cannot open load file" "No such file or directory" "persist")

should be fixed as per above.

i'll push the changes now.

sorry for all the hassle, and let me know if there are any remaining issues.

riscy commented:
> another is that my checking capabilities are clearly inferior to yours, for example regarding autoloads.

It's no problem and thank you for your friendly attitude about it - I think this is one of the challenges of developing elisp; it's sometimes hard to know if code will work well on someone else's system.  Here's the latest on my end --

`lem-api.el` with checkdoc 0.6.2:
```
lem-api.el:187: Lisp symbol `fedi-package-prefix' should appear in quotes
```

`lem-api.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- lem-api.el#L140: Top-level `setq` should usually be replaced by `defvar`
- lem-api.el#L142: Top-level `setq` should usually be replaced by `defvar`
- lem-api.el#L145: Top-level `setq` should usually be replaced by `defvar`
```

`lem-post.el` with byte-compile using Emacs 28.1:
```
lem-post.el:81:1: Warning: defface for `lem-post-community-face' fails to
    specify containing group
In end of data:
lem-post.el:169:24: Warning: the function `lem-list-communities' is not known
    to be defined.
lem-post.el:150:25: Warning: the function `lem-create-post' is not known to be
    defined.
lem-post.el:147:27: Warning: the function `lem-create-comment' is not known to
    be defined.
```

`lem-ui.el` with byte-compile using Emacs 28.1:
```
lem-ui.el:1917:10: Warning: the function `lem-send-private-message' is not
    known to be defined.
lem-ui.el:1759:23: Warning: the function `lem-like-comment' is not known to be
    defined.
lem-ui.el:1758:25: Warning: the function `lem-like-post' is not known to be
    defined.
lem-ui.el:1473:31: Warning: the function `lem-delete-post' is not known to be
    defined.
lem-ui.el:1468:34: Warning: the function `lem-delete-comment' is not known to
    be defined.
lem-ui.el:1450:28: Warning: the function `lem-edit-comment' is not known to be
    defined.
lem-ui.el:1427:6: Warning: the function `lem-mark-private-message-read' is not
    known to be defined.
lem-ui.el:1406:29: Warning: the function `lem-get-private-messages' is not
    known to be defined.
lem-ui.el:1385:21: Warning: the function `lem-get-mentions' is not known to be
    defined.
lem-ui.el:1379:4: Warning: the function `lem-mark-all-read' is not known to be
    defined.
lem-ui.el:1374:6: Warning: the function `lem-mark-comment-reply-read' is not
    known to be defined.
lem-ui.el:1356:20: Warning: the function `lem-get-replies' is not known to be
    defined.
lem-ui.el:1150:21: Warning: the function `lem-follow-community' is not known
    to be defined.
lem-ui.el:1133:17: Warning: the function `lem-list-communities' is not known
    to be defined.
lem-ui.el:1085:29: Warning: the function `lem-save-comment' is not known to be
    defined.
lem-ui.el:1079:29: Warning: the function `lem-save-post' is not known to be
    defined.
lem-ui.el:711:17: Warning: the function `lem-get-community' is not known to be
    defined.
lem-ui.el:678:22: Warning: the function `lem-get-post' is not known to be
    defined.
lem-ui.el:649:24: Warning: the function `lem-resolve-object' is not known to
    be defined.
lem-ui.el:594:21: Warning: the function `lem-search' is not known to be
    defined.
lem-ui.el:348:18: Warning: the function `lem-get-posts' is not known to be
    defined.
lem-ui.el:347:21: Warning: the function `lem-get-instance' is not known to be
    defined.
```

`lem-ui.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- lem-ui.el#L1806: You can use `not` or `null`
```

`lem.el` with byte-compile using Emacs 28.1:
```
lem.el:229:6: Warning: the function `emojify-mode' is not known to be defined.
lem.el:210:17: Warning: the function `lem-get-site' is not known to be
    defined.
lem.el:199:33: Warning: the function `lem-login' is not known to be defined.
```
Sometimes when folks want to optionally support a package (like emojify-mode?) they use a `declare-function` form.

`lem.el` with checkdoc 0.6.2:
```
lem.el:113: First sentence should end with punctuation
```

mooseyboots commented:
all of those not known to be defineds are a result of my `lem-define-request` macro. the file is being required, so all functions in `lem-api.el` should be available. it's a flycheck issue, i'm asking them for further info.

otherwise i have to add a `declare-function` call under every request i define, which is a chore the macro was designed to avoid.

the first two issues you mention i answered in my previous comment.

riscy commented:
> the first two issues you mention i answered in my previous comment.

Apologies for missing your response.  Top-level setq and equivalent should still (in general) be avoided.  Is there any way around this?
```elisp
(defvar fedi-http--api-version)
(setq fedi-http--api-version "v3")
```
The double-hyphens indicate that this is an internal, private variable; part of the fedi package.  Also, what if the user has another package that *also* wants to set this variable?  Who wins?  If you must be using v3 of the API, I'd recommend something an assert or a warning.  You could avoid both the `defvar` and the `setq` with:
```elisp
(if (not (string= fedi-http--api-version "v3"))
    (error "lem requires fedi-http--api-version v3"))
```
Other kinds of top-level `setq`s probably indicate things you should be asking a user to do explicitly in their config, in case there is a race between two packages trying to use the same variable.

> do you perhaps have an idea of how to ensure functions created with a macro are available in compiled code?

I tried a couple things and it looks like the compiler (at least in Emacs 28.1) is not able to figure out what the runtime state would be after those macros have been created.  This is the first time I've seen anything like this!  What seems to work is to put the definition of the macro `lem-def-request` in its own file, and then define all of your different lem functions using that macro in a separate file.  As a really simple example, I created this file:

```elisp
;;; lem-api-2.el --- Hey

(require 'lem-api)

(lem-def-request "post" "delete-post" "post/delete"
  (post-id deleted)
  ""
  (post-id)
  `(("deleted" . ,deleted)))

(provide 'lem-api-2)
;;; lem-api-2.el ends here
```
And now if I `require lem-api-2` in lem-ui.el, I don't get any byte-compile warnings about the non-existence of the delete-post function.  Another option (that I think you have tried based on code I've looked at) is to make use of declare-function, but this requires additional bookkeeping.

mooseyboots commented:
thanks for the setq feedback, i need to think about `fedi` a little more then. it's meant for package writers, but i need to re-think the variables.

& thanks for finding a solution re the macro! i read about the separate file bizzo in the manual when looking for a solution, but figured i already had a separate file. now i see we need a separate file for the separate file. can update the repo when i have the chance.

i was still a little confused also whether/how `declare-function` would work or not. best if i can avoid needing it.

mooseyboots commented:
added separate `lem-request.el` file, and `lem.el` now byte-compiles cleanly for me.

i made a tiny adjustment to `fedi.el`, so i could remove those setqs. lem can have the variables and send them as args to fedi.

thanks again for the help.

mooseyboots commented:
@riscy i think this should be good to go now? the last two issues should be resolved. unless you've since found any further issues.

riscy commented:
I apologize for that, it looks like I missed this one when going through the PRs yesterday.  This looks good to go!
